File: src/plugins/sandbox.ts
/**
 * 플러그인 샌드박스
 * 플러그인을 격리된 환경에서 실행
 */

import { Worker, isMainThread, parentPort, workerData } from 'worker_threads';
import { EventEmitter } from 'events';
import { createHash } from 'crypto';
import { join } from 'path';
import {
  PluginDescriptor,
  PluginSandboxInfo,
  PluginPermission
} from './types.js';

/**
 * 샌드박스 설정
 */
export interface SandboxConfig {
  enabled: boolean;
  memoryLimit: number;      // bytes
  cpuLimit: number;         // percentage
  networkAllowed: boolean;
  fileSystemAccess: 'none' | 'readonly' | 'readwrite';
}

/**
 * 샌드박스 환경 정보
 */
export interface SandboxEnvironment {
  id: string;
  pluginId: string;
  worker?: Worker;
  isolationLevel: 'none' | 'basic' | 'strict';
  resourceLimits: {
    memory: number;
    cpu: number;
    files: number;
  };
  allowedAPIs: string[];
  startTime: Date;
  lastActivity: Date;
}

/**
 * 플러그인 샌드박스 매니저
 */
export class PluginSandbox extends EventEmitter {
  private environments = new Map<string, SandboxEnvironment>();
  private resourceMonitor?: NodeJS.Timeout;

  constructor(
    private config: SandboxConfig
  ) {
    super();
    
    if (this.config.enabled) {
      this.startResourceMonitoring();
    }
  }

  /**
   * 샌드박스 환경 생성
   */
  async createEnvironment(descriptor: PluginDescriptor): Promise<PluginSandboxInfo> {
    const pluginId = descriptor.id;
    const permissions = descriptor.manifest.permissions || [];
    
    console.log(`[PluginSandbox] Creating environment for plugin: ${pluginId}`);

    // 격리 레벨 결정
    const isolationLevel = this.determineIsolationLevel(permissions);
    
    // 리소스 제한 설정
    const resourceLimits = {
      memory: this.config.memoryLimit,
      cpu: this.config.cpuLimit,
      files: this.getFileLimit(permissions)
    };

    // 허용된 API 목록 생성
    const allowedAPIs = this.generateAllowedAPIs(permissions);

    const environment: SandboxEnvironment = {
      id: this.generateEnvironmentId(pluginId),
      pluginId,
      isolationLevel,
      resourceLimits,
      allowedAPIs,
      startTime: new Date(),
      lastActivity: new Date()
    };

    // 워커 스레드 생성 (strict 격리 레벨인 경우)
    if (isolationLevel === 'strict') {
      environment.worker = await this.createWorkerThread(descriptor, environment);
    }

    this.environments.set(pluginId, environment);

    const sandboxInfo: PluginSandboxInfo = {
      pid: environment.worker?.threadId,
      isolationLevel,
      resourceLimits,
      allowedAPIs
    };

    this.emit('environment.created', { pluginId, environment: sandboxInfo });
    console.log(`[PluginSandbox] Environment created for ${pluginId} with ${isolationLevel} isolation`);

    return sandboxInfo;
  }

  /**
   * 샌드박스 환경 해제
   */
  async destroyEnvironment(pluginId: string): Promise<void> {
    const environment = this.environments.get(pluginId);
    if (!environment) {
      return;
    }

    console.log(`[PluginSandbox] Destroying environment for plugin: ${pluginId}`);

    // 워커 스레드 종료
    if (environment.worker) {
      await environment.worker.terminate();
    }

    this.environments.delete(pluginId);
    this.emit('environment.destroyed', { pluginId });

    console.log(`[PluginSandbox] Environment destroyed for plugin: ${pluginId}`);
  }

  /**
   * 격리 레벨 결정
   */
  private determineIsolationLevel(permissions: PluginPermission[]): 'none' | 'basic' | 'strict' {
    // 위험한 권한들
    const dangerousPermissions = [
      PluginPermission.WRITE_FILES,
      PluginPermission.NETWORK_ACCESS,
      PluginPermission.SYSTEM_INFO,
      PluginPermission.DATABASE_WRITE,
      PluginPermission.SECURITY
    ];

    // 중간 위험 권한들
    const moderatePermissions = [
      PluginPermission.READ_FILES,
      PluginPermission.DATABASE_READ,
      PluginPermission.USER_DATA
    ];

    const hasDangerousPerms = permissions.some(p => dangerousPermissions.includes(p));
    const hasModeratePerms = permissions.some(p => moderatePermissions.includes(p));

    if (hasDangerousPerms) {
      return 'strict';
    } else if (hasModeratePerms) {
      return 'basic';
    } else {
      return 'none';
    }
  }

  /**
   * 파일 접근 제한 수 계산
   */
  private getFileLimit(permissions: PluginPermission[]): number {
    if (permissions.includes(PluginPermission.WRITE_FILES)) {
      return 1000; // 많은 파일 접근 허용
    } else if (permissions.includes(PluginPermission.READ_FILES)) {
      return 500;  // 제한된 파일 접근
    } else {
      return 0;    // 파일 접근 없음
    }
  }

  /**
   * 허용된 API 목록 생성
   */
  private generateAllowedAPIs(permissions: PluginPermission[]): string[] {
    const allowedAPIs: string[] = [];

    // 기본 API는 항상 허용
    allowedAPIs.push('logger', 'events', 'storage', 'communication');

    // 권한에 따른 API 추가
    if (permissions.includes(PluginPermission.READ_FILES) || 
        permissions.includes(PluginPermission.WRITE_FILES)) {
      allowedAPIs.push('fs');
    }

    if (permissions.includes(PluginPermission.NETWORK_ACCESS)) {
      allowedAPIs.push('http');
    }

    if (permissions.includes(PluginPermission.DATABASE_READ) || 
        permissions.includes(PluginPermission.DATABASE_WRITE)) {
      allowedAPIs.push('database');
    }

    if (permissions.includes(PluginPermission.MCP_TOOLS)) {
      allowedAPIs.push('mcp');
    }

    if (permissions.includes(PluginPermission.NOTIFICATIONS)) {
      allowedAPIs.push('notifications');
    }

    return allowedAPIs;
  }

  /**
   * 워커 스레드 생성
   */
  private async createWorkerThread(
    descriptor: PluginDescriptor, 
    environment: SandboxEnvironment
  ): Promise<Worker> {
    const workerScript = join(__dirname, 'sandbox-worker.js');
    
    const worker = new Worker(workerScript, {
      workerData: {
        pluginPath: descriptor.path,
        pluginManifest: descriptor.manifest,
        environment: {
          id: environment.id,
          resourceLimits: environment.resourceLimits,
          allowedAPIs: environment.allowedAPIs
        }
      },
      resourceLimits: {
        maxOldGenerationSizeMb: Math.floor(environment.resourceLimits.memory / 1024 / 1024),
        maxYoungGenerationSizeMb: Math.floor(environment.resourceLimits.memory / 1024 / 1024 / 2)
      }
    });

    // 워커 이벤트 처리
    worker.on('message', (message) => {
      this.handleWorkerMessage(descriptor.id, message);
    });

    worker.on('error', (error) => {
      console.error(`[PluginSandbox] Worker error for ${descriptor.id}:`, error);
      this.emit('environment.error', { pluginId: descriptor.id, error });
    });

    worker.on('exit', (code) => {
      if (code !== 0) {
        console.error(`[PluginSandbox] Worker exited with code ${code} for ${descriptor.id}`);
      }
      this.emit('environment.exited', { pluginId: descriptor.id, code });
    });

    return worker;
  }

  /**
   * 워커 메시지 처리
   */
  private handleWorkerMessage(pluginId: string, message: any): void {
    const environment = this.environments.get(pluginId);
    if (!environment) {
      return;
    }

    environment.lastActivity = new Date();

    switch (message.type) {
      case 'api_call':
        this.handleAPICall(pluginId, message);
        break;
      
      case 'log':
        console.log(`[Plugin:${pluginId}] ${message.level}: ${message.message}`);
        break;
      
      case 'metric':
        this.emit('environment.metric', { pluginId, metric: message.data });
        break;
      
      default:
        console.warn(`[PluginSandbox] Unknown message type from ${pluginId}:`, message.type);
    }
  }

  /**
   * API 호출 처리
   */
  private handleAPICall(pluginId: string, message: any): void {
    const environment = this.environments.get(pluginId);
    if (!environment) {
      return;
    }

    const { api, method, args, callId } = message;

    // API 접근 권한 확인
    if (!environment.allowedAPIs.includes(api)) {
      this.sendToWorker(pluginId, {
        type: 'api_response',
        callId,
        error: `API '${api}' not allowed for this plugin`
      });
      return;
    }

    // API 호출 처리 (실제 구현 필요)
    this.processAPICall(pluginId, api, method, args)
      .then(result => {
        this.sendToWorker(pluginId, {
          type: 'api_response',
          callId,
          result
        });
      })
      .catch(error => {
        this.sendToWorker(pluginId, {
          type: 'api_response',
          callId,
          error: error.message
        });
      });
  }

  /**
   * API 호출 실제 처리
   */
  private async processAPICall(pluginId: string, api: string, method: string, args: any[]): Promise<any> {
    // 실제 API 호출 처리 로직
    // 각 API별로 적절한 핸들러로 전달
    
    switch (api) {
      case 'logger':
        return this.handleLoggerAPI(pluginId, method, args);
      
      case 'storage':
        return this.handleStorageAPI(pluginId, method, args);
      
      case 'fs':
        return this.handleFileSystemAPI(pluginId, method, args);
      
      case 'http':
        return this.handleHTTPAPI(pluginId, method, args);
      
      default:
        throw new Error(`Unknown API: ${api}`);
    }
  }

  /**
   * 로거 API 처리
   */
  private async handleLoggerAPI(pluginId: string, method: string, args: any[]): Promise<any> {
    const [level, message, meta] = args;
    console.log(`[Plugin:${pluginId}] [${level.toUpperCase()}] ${message}`, meta || '');
    return true;
  }

  /**
   * 스토리지 API 처리
   */
  private async handleStorageAPI(pluginId: string, method: string, args: any[]): Promise<any> {
    // 실제 스토리지 구현과 연결
    throw new Error('Storage API not implemented yet');
  }

  /**
   * 파일 시스템 API 처리
   */
  private async handleFileSystemAPI(pluginId: string, method: string, args: any[]): Promise<any> {
    // 실제 파일 시스템 구현과 연결
    throw new Error('FileSystem API not implemented yet');
  }

  /**
   * HTTP API 처리
   */
  private async handleHTTPAPI(pluginId: string, method: string, args: any[]): Promise<any> {
    // 실제 HTTP 클라이언트 구현과 연결
    throw new Error('HTTP API not implemented yet');
  }

  /**
   * 워커에 메시지 전송
   */
  private sendToWorker(pluginId: string, message: any): void {
    const environment = this.environments.get(pluginId);
    if (environment?.worker) {
      environment.worker.postMessage(message);
    }
  }

  /**
   * 리소스 모니터링 시작
   */
  private startResourceMonitoring(): void {
    this.resourceMonitor = setInterval(() => {
      this.monitorResources();
    }, 5000); // 5초마다 모니터링
  }

  /**
   * 리소스 모니터링
   */
  private monitorResources(): void {
    for (const [pluginId, environment] of this.environments) {
      try {
        this.checkResourceUsage(pluginId, environment);
      } catch (error) {
        console.error(`[PluginSandbox] Error monitoring ${pluginId}:`, error);
      }
    }
  }

  /**
   * 리소스 사용량 체크
   */
  private checkResourceUsage(pluginId: string, environment: SandboxEnvironment): void {
    if (!environment.worker) {
      return;
    }

    // 메모리 사용량 체크 (워커 스레드의 경우 정확한 측정이 어려움)
    const memoryUsage = process.memoryUsage();
    const estimatedPluginMemory = memoryUsage.heapUsed / this.environments.size;

    if (estimatedPluginMemory > environment.resourceLimits.memory) {
      console.warn(`[PluginSandbox] Plugin ${pluginId} exceeding memory limit`);
      this.emit('environment.resource.exceeded', {
        pluginId,
        resource: 'memory',
        usage: estimatedPluginMemory,
        limit: environment.resourceLimits.memory
      });
    }

    // 비활성 플러그인 감지 (30분 이상 비활성)
    const inactiveTime = Date.now() - environment.lastActivity.getTime();
    if (inactiveTime > 30 * 60 * 1000) {
      console.warn(`[PluginSandbox] Plugin ${pluginId} has been inactive for ${Math.floor(inactiveTime / 60000)} minutes`);
      this.emit('environment.inactive', { pluginId, inactiveTime });
    }
  }

  /**
   * 환경 ID 생성
   */
  private generateEnvironmentId(pluginId: string): string {
    const timestamp = Date.now().toString();
    const hash = createHash('sha256').update(pluginId + timestamp).digest('hex');
    return `env_${hash.substring(0, 8)}`;
  }

  /**
   * 환경 정보 조회
   */
  getEnvironment(pluginId: string): SandboxEnvironment | null {
    return this.environments.get(pluginId) || null;
  }

  /**
   * 모든 환경 정보 조회
   */
  getAllEnvironments(): SandboxEnvironment[] {
    return Array.from(this.environments.values());
  }

  /**
   * 정리
   */
  async dispose(): Promise<void> {
    console.log('[PluginSandbox] Disposing...');

    // 리소스 모니터링 중지
    if (this.resourceMonitor) {
      clearInterval(this.resourceMonitor);
    }

    // 모든 환경 해제
    const pluginIds = Array.from(this.environments.keys());
    for (const pluginId of pluginIds) {
      await this.destroyEnvironment(pluginId);
    }

    console.log('[PluginSandbox] Disposed');
  }
}

File: python/src/devflow_monitor/plugins/sandbox.py
"""
Plugin Sandbox.

This module provides isolated execution environments for plugins using
multiprocessing. Implements resource limits, security restrictions,
and monitored execution.
"""

import asyncio
import hashlib
import multiprocessing
import os
import time
from datetime import datetime
from multiprocessing import Process, Queue
from queue import Empty
from typing import Any, Callable, Optional

from .types import (
    IsolationLevel,
    PluginDescriptor,
    PluginPermission,
    PluginResourceLimits,
    PluginSandboxInfo,
)


class SandboxConfig:
    """Sandbox configuration."""

    def __init__(
        self,
        enabled: bool = True,
        memory_limit: int = 536870912,  # 512MB
        cpu_limit: int = 80,  # percentage
        network_allowed: bool = True,
        filesystem_access: str = "readonly",
    ):
        """
        Initialize sandbox configuration.

        Args:
            enabled: Whether sandboxing is enabled.
            memory_limit: Memory limit in bytes.
            cpu_limit: CPU limit percentage.
            network_allowed: Whether network access is allowed.
            filesystem_access: Filesystem access level (none/readonly/readwrite).
        """
        self.enabled = enabled
        self.memory_limit = memory_limit
        self.cpu_limit = cpu_limit
        self.network_allowed = network_allowed
        self.filesystem_access = filesystem_access


class SandboxEnvironment:
    """Sandbox environment for a plugin."""

    def __init__(
        self,
        env_id: str,
        plugin_id: str,
        isolation_level: IsolationLevel,
        resource_limits: PluginResourceLimits,
        allowed_apis: list[str],
    ):
        """
        Initialize sandbox environment.

        Args:
            env_id: Environment unique ID.
            plugin_id: Plugin ID.
            isolation_level: Isolation level.
            resource_limits: Resource limits.
            allowed_apis: List of allowed APIs.
        """
        self.id = env_id
        self.plugin_id = plugin_id
        self.isolation_level = isolation_level
        self.resource_limits = resource_limits
        self.allowed_apis = allowed_apis
        self.start_time = datetime.utcnow()
        self.last_activity = datetime.utcnow()
        self.process: Optional[Process] = None
        self.input_queue: Optional[Queue] = None
        self.output_queue: Optional[Queue] = None


def _sandbox_worker(
    plugin_path: str,
    plugin_manifest: dict[str, Any],
    environment: dict[str, Any],
    input_queue: Queue,
    output_queue: Queue,
) -> None:
    """
    Sandbox worker process function.

    Runs in a separate process with restricted capabilities.

    Args:
        plugin_path: Path to the plugin.
        plugin_manifest: Plugin manifest data.
        environment: Environment configuration.
        input_queue: Queue for receiving commands.
        output_queue: Queue for sending responses.
    """
    try:
        # Apply resource limits (platform specific)
        try:
            import resource

            # Set memory limit
            memory_limit = environment.get("resource_limits", {}).get(
                "memory", 536870912
            )
            resource.setrlimit(
                resource.RLIMIT_AS, (memory_limit, memory_limit)
            )

            # Set CPU time limit
            cpu_limit = environment.get("resource_limits", {}).get("timeout", 30)
            resource.setrlimit(resource.RLIMIT_CPU, (cpu_limit, cpu_limit))

        except (ImportError, OSError):
            # Resource module not available on all platforms
            pass

        # Restricted globals for exec
        restricted_globals = {
            "__builtins__": {
                "print": print,
                "len": len,
                "range": range,
                "str": str,
                "int": int,
                "float": float,
                "bool": bool,
                "list": list,
                "dict": dict,
                "tuple": tuple,
                "set": set,
                "True": True,
                "False": False,
                "None": None,
                "Exception": Exception,
                "ValueError": ValueError,
                "TypeError": TypeError,
                "RuntimeError": RuntimeError,
            }
        }

        # Add allowed APIs based on permissions
        allowed_apis = environment.get("allowed_apis", [])

        if "http" in allowed_apis:
            import urllib.request

            restricted_globals["__builtins__"]["urllib"] = urllib

        if "json" in allowed_apis:
            import json

            restricted_globals["json"] = json

        while True:
            try:
                # Wait for command from main process
                command = input_queue.get(timeout=1)

                if command is None or command.get("type") == "shutdown":
                    break

                command_type = command.get("type")
                call_id = command.get("call_id")

                if command_type == "execute":
                    code = command.get("code", "")
                    local_vars: dict[str, Any] = {}

                    try:
                        exec(code, restricted_globals, local_vars)
                        output_queue.put({
                            "type": "result",
                            "call_id": call_id,
                            "result": local_vars.get("result"),
                        })
                    except Exception as e:
                        output_queue.put({
                            "type": "error",
                            "call_id": call_id,
                            "error": str(e),
                        })

                elif command_type == "api_call":
                    api = command.get("api")
                    method = command.get("method")
                    args = command.get("args", [])

                    # Check if API is allowed
                    if api not in allowed_apis:
                        output_queue.put({
                            "type": "error",
                            "call_id": call_id,
                            "error": f"API '{api}' not allowed for this plugin",
                        })
                        continue

                    # Handle API calls (simplified)
                    output_queue.put({
                        "type": "api_response",
                        "call_id": call_id,
                        "result": None,
                        "message": f"API {api}.{method} called with {args}",
                    })

                elif command_type == "health_check":
                    output_queue.put({
                        "type": "health",
                        "call_id": call_id,
                        "status": "healthy",
                    })

            except Empty:
                continue
            except Exception as e:
                output_queue.put({
                    "type": "error",
                    "error": str(e),
                })

    except Exception as e:
        output_queue.put({
            "type": "fatal_error",
            "error": str(e),
        })


class PluginSandbox:
    """
    Plugin Sandbox Manager.

    Provides isolated execution environments for plugins using multiprocessing.
    Implements resource limits, security restrictions, and API access control.

    Args:
        config: Sandbox configuration.

    Example:
        >>> config = SandboxConfig(enabled=True, memory_limit=512*1024*1024)
        >>> sandbox = PluginSandbox(config)
        >>> info = await sandbox.create_environment(descriptor)
        >>> result = await sandbox.execute(plugin_id, "result = 1 + 1")
    """

    def __init__(self, config: SandboxConfig):
        """Initialize the plugin sandbox."""
        self.config = config
        self._environments: dict[str, SandboxEnvironment] = {}
        self._monitor_task: Optional[asyncio.Task] = None
        self._event_handlers: dict[str, list[Callable[..., Any]]] = {}

        if self.config.enabled:
            self._start_resource_monitoring()

    def on(self, event: str, handler: Callable[..., Any]) -> None:
        """
        Register an event handler.

        Args:
            event: Event name.
            handler: Event handler function.
        """
        if event not in self._event_handlers:
            self._event_handlers[event] = []
        self._event_handlers[event].append(handler)

    def emit(self, event: str, data: Any) -> None:
        """
        Emit an event to all registered handlers.

        Args:
            event: Event name.
            data: Event data.
        """
        handlers = self._event_handlers.get(event, [])
        for handler in handlers:
            try:
                handler(data)
            except Exception:
                pass

    async def create_environment(
        self, descriptor: PluginDescriptor
    ) -> PluginSandboxInfo:
        """
        Create a sandbox environment for a plugin.

        Args:
            descriptor: Plugin descriptor.

        Returns:
            PluginSandboxInfo with environment details.
        """
        plugin_id = descriptor.id
        permissions = descriptor.manifest.permissions or []

        # Determine isolation level based on permissions
        isolation_level = self._determine_isolation_level(permissions)

        # Set resource limits
        resource_limits = PluginResourceLimits(
            memory=self.config.memory_limit,
            cpu=self.config.cpu_limit,
            files=self._get_file_limit(permissions),
        )

        # Generate allowed APIs based on permissions
        allowed_apis = self._generate_allowed_apis(permissions)

        environment = SandboxEnvironment(
            env_id=self._generate_environment_id(plugin_id),
            plugin_id=plugin_id,
            isolation_level=isolation_level,
            resource_limits=resource_limits,
            allowed_apis=allowed_apis,
        )

        # Create worker process for strict isolation
        if isolation_level == IsolationLevel.STRICT:
            await self._create_worker_process(descriptor, environment)

        self._environments[plugin_id] = environment

        sandbox_info = PluginSandboxInfo(
            pid=environment.process.pid if environment.process else None,
            isolation_level=isolation_level,
            resource_limits=resource_limits,
            allowed_apis=allowed_apis,
        )

        self.emit("environment.created", {"plugin_id": plugin_id, "environment": sandbox_info})
        return sandbox_info

    async def destroy_environment(self, plugin_id: str) -> None:
        """
        Destroy a sandbox environment.

        Args:
            plugin_id: Plugin identifier.
        """
        environment = self._environments.get(plugin_id)
        if not environment:
            return

        # Terminate worker process
        if environment.process and environment.process.is_alive():
            # Send shutdown command
            if environment.input_queue:
                environment.input_queue.put({"type": "shutdown"})

            # Wait for graceful shutdown
            environment.process.join(timeout=5)

            # Force terminate if still running
            if environment.process.is_alive():
                environment.process.terminate()
                environment.process.join(timeout=2)

        del self._environments[plugin_id]
        self.emit("environment.destroyed", {"plugin_id": plugin_id})

    async def execute(
        self,
        plugin_id: str,
        code: str,
        context: Optional[dict[str, Any]] = None,
        timeout: float = 30.0,
    ) -> Any:
        """
        Execute code in a sandbox environment.

        Args:
            plugin_id: Plugin identifier.
            code: Code to execute.
            context: Optional execution context.
            timeout: Execution timeout in seconds.

        Returns:
            Execution result.

        Raises:
            ValueError: If environment not found.
            TimeoutError: If execution times out.
            RuntimeError: If execution fails.
        """
        environment = self._environments.get(plugin_id)
        if not environment:
            raise ValueError(f"Sandbox environment not found for plugin: {plugin_id}")

        environment.last_activity = datetime.utcnow()

        if environment.isolation_level == IsolationLevel.STRICT:
            return await self._execute_in_process(environment, code, context, timeout)
        else:
            return await self._execute_inline(environment, code, context, timeout)

    async def _execute_in_process(
        self,
        environment: SandboxEnvironment,
        code: str,
        context: Optional[dict[str, Any]],
        timeout: float,
    ) -> Any:
        """
        Execute code in a separate process.

        Args:
            environment: Sandbox environment.
            code: Code to execute.
            context: Execution context.
            timeout: Timeout in seconds.

        Returns:
            Execution result.
        """
        if not environment.input_queue or not environment.output_queue:
            raise RuntimeError("Process communication not initialized")

        call_id = hashlib.md5(f"{time.time()}{code}".encode()).hexdigest()[:8]

        # Send execute command
        environment.input_queue.put({
            "type": "execute",
            "call_id": call_id,
            "code": code,
            "context": context,
        })

        # Wait for response
        start_time = time.time()
        while time.time() - start_time < timeout:
            try:
                response = environment.output_queue.get(timeout=0.1)
                if response.get("call_id") == call_id:
                    if response.get("type") == "error":
                        raise RuntimeError(response.get("error"))
                    return response.get("result")
            except Empty:
                continue

        raise TimeoutError(f"Execution timeout ({timeout}s)")

    async def _execute_inline(
        self,
        environment: SandboxEnvironment,
        code: str,
        context: Optional[dict[str, Any]],
        timeout: float,
    ) -> Any:
        """
        Execute code inline with basic restrictions.

        Args:
            environment: Sandbox environment.
            code: Code to execute.
            context: Execution context.
            timeout: Timeout in seconds.

        Returns:
            Execution result.
        """
        # Create restricted execution environment
        restricted_globals = {
            "__builtins__": {
                "print": print,
                "len": len,
                "range": range,
                "str": str,
                "int": int,
                "float": float,
                "bool": bool,
                "list": list,
                "dict": dict,
                "tuple": tuple,
                "set": set,
                "True": True,
                "False": False,
                "None": None,
                "Exception": Exception,
            }
        }

        if context:
            restricted_globals.update(context)

        local_vars: dict[str, Any] = {}

        try:
            exec(code, restricted_globals, local_vars)
            return local_vars.get("result")
        except Exception as e:
            raise RuntimeError(f"Execution failed: {e}")

    async def _create_worker_process(
        self,
        descriptor: PluginDescriptor,
        environment: SandboxEnvironment,
    ) -> None:
        """
        Create a worker process for strict isolation.

        Args:
            descriptor: Plugin descriptor.
            environment: Sandbox environment.
        """
        environment.input_queue = multiprocessing.Queue()
        environment.output_queue = multiprocessing.Queue()

        environment.process = Process(
            target=_sandbox_worker,
            args=(
                descriptor.path,
                descriptor.manifest.model_dump(),
                {
                    "id": environment.id,
                    "resource_limits": environment.resource_limits.model_dump(),
                    "allowed_apis": environment.allowed_apis,
                },
                environment.input_queue,
                environment.output_queue,
            ),
        )

        environment.process.start()

        # Wait for process to be ready
        await asyncio.sleep(0.1)

    def _determine_isolation_level(
        self, permissions: list[PluginPermission]
    ) -> IsolationLevel:
        """
        Determine isolation level based on permissions.

        Args:
            permissions: List of required permissions.

        Returns:
            Appropriate isolation level.
        """
        # Dangerous permissions require strict isolation
        dangerous_permissions = [
            PluginPermission.FILE_WRITE,
            PluginPermission.NETWORK_ACCESS,
            PluginPermission.SYSTEM_INFO,
            PluginPermission.DATABASE_WRITE,
            PluginPermission.SECURITY,
        ]

        # Moderate permissions require basic isolation
        moderate_permissions = [
            PluginPermission.FILE_READ,
            PluginPermission.DATABASE_READ,
            PluginPermission.USER_DATA,
        ]

        if isinstance(permissions, list):
            perm_set = set(permissions)
        else:
            perm_set = set()

        if perm_set & set(dangerous_permissions):
            return IsolationLevel.STRICT
        elif perm_set & set(moderate_permissions):
            return IsolationLevel.BASIC
        else:
            return IsolationLevel.NONE

    def _get_file_limit(self, permissions: list[PluginPermission]) -> int:
        """
        Get file handle limit based on permissions.

        Args:
            permissions: List of permissions.

        Returns:
            Maximum file handles.
        """
        if PluginPermission.FILE_WRITE in permissions:
            return 1000
        elif PluginPermission.FILE_READ in permissions:
            return 500
        else:
            return 0

    def _generate_allowed_apis(
        self, permissions: list[PluginPermission]
    ) -> list[str]:
        """
        Generate list of allowed APIs based on permissions.

        Args:
            permissions: List of permissions.

        Returns:
            List of allowed API names.
        """
        allowed_apis = ["logger", "events", "storage", "communication"]

        perm_set = set(permissions) if isinstance(permissions, list) else set()

        if (
            PluginPermission.FILE_READ in perm_set
            or PluginPermission.FILE_WRITE in perm_set
        ):
            allowed_apis.append("fs")

        if PluginPermission.NETWORK_ACCESS in perm_set:
            allowed_apis.append("http")
            allowed_apis.append("json")

        if (
            PluginPermission.DATABASE_READ in perm_set
            or PluginPermission.DATABASE_WRITE in perm_set
        ):
            allowed_apis.append("database")

        if PluginPermission.MCP_TOOLS in perm_set:
            allowed_apis.append("mcp")

        if PluginPermission.NOTIFICATIONS in perm_set:
            allowed_apis.append("notifications")

        return allowed_apis

    def _generate_environment_id(self, plugin_id: str) -> str:
        """
        Generate unique environment ID.

        Args:
            plugin_id: Plugin identifier.

        Returns:
            Unique environment ID.
        """
        timestamp = str(time.time())
        hash_input = f"{plugin_id}{timestamp}"
        return f"env_{hashlib.sha256(hash_input.encode()).hexdigest()[:8]}"

    def get_environment(self, plugin_id: str) -> Optional[SandboxEnvironment]:
        """
        Get sandbox environment for a plugin.

        Args:
            plugin_id: Plugin identifier.

        Returns:
            SandboxEnvironment if exists, None otherwise.
        """
        return self._environments.get(plugin_id)

    def get_all_environments(self) -> list[SandboxEnvironment]:
        """
        Get all sandbox environments.

        Returns:
            List of all environments.
        """
        return list(self._environments.values())

    def _start_resource_monitoring(self) -> None:
        """Start resource monitoring task."""
        async def monitor_loop():
            while True:
                await asyncio.sleep(5)  # Check every 5 seconds
                self._monitor_resources()

        try:
            loop = asyncio.get_event_loop()
            self._monitor_task = loop.create_task(monitor_loop())
        except RuntimeError:
            # No event loop running yet
            pass

    def _monitor_resources(self) -> None:
        """Monitor resource usage of all environments."""
        for plugin_id, environment in list(self._environments.items()):
            try:
                self._check_resource_usage(plugin_id, environment)
            except Exception as e:
                print(f"[PluginSandbox] Error monitoring {plugin_id}: {e}")

    def _check_resource_usage(
        self, plugin_id: str, environment: SandboxEnvironment
    ) -> None:
        """
        Check resource usage for an environment.

        Args:
            plugin_id: Plugin identifier.
            environment: Sandbox environment.
        """
        if not environment.process:
            return

        # Check if process is still alive
        if not environment.process.is_alive():
            self.emit("environment.exited", {
                "plugin_id": plugin_id,
                "exit_code": environment.process.exitcode,
            })
            return

        # Check for inactive environments (30 minutes)
        inactive_time = (
            datetime.utcnow() - environment.last_activity
        ).total_seconds()
        if inactive_time > 30 * 60:
            self.emit("environment.inactive", {
                "plugin_id": plugin_id,
                "inactive_time": inactive_time,
            })

    async def dispose(self) -> None:
        """Dispose of the sandbox manager and all environments."""
        # Cancel monitor task
        if self._monitor_task:
            self._monitor_task.cancel()
            try:
                await self._monitor_task
            except asyncio.CancelledError:
                pass

        # Destroy all environments
        for plugin_id in list(self._environments.keys()):
            await self.destroy_environment(plugin_id)
